doc/contribute/contributor_expectations.rst

```diff
@@ -87,27 +87,40 @@ Changes which require an RFC proposal include:
 Maintainers have the discretion to request that contributors create an RFC for
 PRs that are too large or complicated.
 
+.. _pr_requirements:
+
 PR Requirements
 ***************
 
 - Each commit in the PR must provide a commit message following the
   :ref:`commit-guidelines`.
 
+- No fixup or merge commits are allowed, see :ref:`Contribution workflow` for
+  more information.
+
 - The PR description must include a summary of the changes and their rationales.
 
 - All files in the PR must comply with :ref:`Licensing
   Requirements<licensing_requirements>`.
 
-- Follow the Zephyr :ref:`coding_style` and :ref:`coding_guidelines`.
+- The code must follow the Zephyr :ref:`coding_style` and :ref:`coding_guidelines`.
 
-- PRs must pass all CI checks. This is a requirement to merge the PR.
+- The PR must pass all CI checks, as described in :ref:`merge_criteria`.
   Contributors may mark a PR as draft and explicitly request reviewers to
   provide early feedback, even with failing CI checks.
 
-- When breaking a PR into multiple commits, each commit must build cleanly. The
+- When breaking up a PR into multiple commits, each commit must build cleanly. The
   CI system does not enforce this policy, so it is the PR author's
   responsibility to verify.
 
+- Commits in a pull request must represent distinct, reviewable, logical units
+  of change that retain bisectibility. Unrelated changes should not be present
+  in the same commit. Intermediary changes or non-final artifacts that occur in
+  the git history during development of a pull request must be squashed out of
+  these fundamental commits whenever bisectability allows it. This includes, but
+  is not limited to, renaming files that have been introduced in earlier
+  commits, or rewriting newly added code.
```

Comment:
The first sentence is more then sufficient to make the point clear.

> Commits in a pull request must represent distinct, reviewable, logical units
  of change that retain bisectibility.

The remaining too restrictive If reviewers will not enforce this commit by commit project will have a double standard problem. The way it is written blocks collaboration and no one can pick a PR and continue and it not allow more then one person per PR when they want to have distinguishing contributions.

CC: @kartben

Reply:
You keep the original submitter's sign off and add yours, with your changes applied, as part of the same commit

Reply:
@nandojve I think that what you state is not accurate. We have a full section on what you can or cannot do with other people's code when collaborating: https://docs.zephyrproject.org/latest/project/modifying_contributions.html.
Nothing in the text prevents you from using the rules above to collaborate on PRs. But even more importantly, I feel that we are taking these rules a bit too far. This is not a legally binding contract and there will always be case by case decisions made on the basis of what is written. The text is designed to convey a general idea, not to cover every single specific case.

Reply:
@nandojve there's one more factor I'd like to mention here, which is collaboration. If several people are working on a PR together, then they can decide among themselves how to split and attribute the commits' authorship. On the other hand, if someone is taking over a PR from a contributor no longer involved, then they have a set of rules in the link I pasted above.

Reply:
@nordicjm, 

Distinguishing contributions is not so much about sign off, it is more about copyrights.

@carlescufi,

This opens the doors to more complex subjects, for instance, this part: _This includes, but   is not limited to, renaming files that have been introduced in earlier commits, or rewriting newly added code._ You must keep original contributions "cherry-pick" and fix with new commits. It is not allowed to squash then. This means that, for instance, a file rename should go as is on a cherry-pick commit, if is that the case. Otherwise we break the principle of _distinguishing contributions_. The project must keep copyrights from author and no one is allowed to overtake it. The "cherry-pick" will have a sign off from new author because it is a cherry-pick and required that attribution to commit on the tree.

Usually, people will learn how to do from here and makes mistakes there. This is why I'm mention that this rule is too strict and should be dropped on my opinion.

> @nandojve there's one more factor I'd like to mention here, which is collaboration. If several people are working on a PR together, then they can decide among themselves how to split and attribute the commits' authorship.

Yes, but people on the project start to say that authors should squash their commits and just add a sign off because this is not on the interest of the project. This means that the above is on the wrong way already.

> On the other hand, if someone is taking over a PR from a contributor no longer involved, then they have a set of rules in the link I pasted above.

Yes, and now we are established another set of rules that interpose. How do you guarantee that people will make this very clear? Should we need to update that documentation and say that everything here do not apply?

Well, this is why I'm a little afraid about this particular change:

1- this change was triggered at "12 pm" when very low people is looking on the project.
2- interpose of rules may lead to misinterpretation and double standards or mistakes that could imply law.
3- no broad discussion. This should go over an RFC and make fully transparent inviting the community to take part of this change on the mail list.
4- not everyone know their rights or understand the nuances about this changes. This means that I would expect that Linux Foundation should be the Assignee here and on all rules changes. Change on rules should pass by governance to avoid a small group of people promote too big changes.

I believe have the rules documented is a big step to be transparent and set expectations but this part is too strict.

Reply:
> Distinguishing contributions is not so much about sign off, it is more about copyrights.

The copyright is apache, there is nothing else needed. If someone wants to add their name to the copyright list at the top of the file, they can do that. A commit does not imply copyright

Reply:
Again, you don't know what you are talking about.

Reply:
I really wanted to avoid this, but I see it's unavoidable.

@nandojve 

>This opens the doors to more complex subjects, for instance, this part: This includes, but is not limited to, renaming files that have been introduced in earlier commits, or rewriting newly added code. You must keep original contributions "cherry-pick" and fix with new commits. It is not allowed to squash then. This means that, for instance, a file rename should go as is on a cherry-pick commit, if is that the case. Otherwise we break the principle of distinguishing contributions. The project must keep copyrights from author and no one is allowed to overtake it. The "cherry-pick" will have a sign off from new author because it is a cherry-pick and required that attribution to commit on the tree.

>Yes, but people on the project start to say that authors should squash their commits and just add a sign off because this is not on the interest of the project. This means that the above is on the wrong way already.

Again, these are guidelines, not a precise rulebook nor a contract. Ultimately engineers should an, in fact, must, decide together what makes sense in every different case. Of course that cherry-picking a commit from someone else may collide with the guidance in this paragraph, I agree. But this is something we cannot fix without having multiple pages of detailed rules that specify exactly what to do in each scenario.
The reason that we added the text that you suggest we remove is that we had a real scenario where a contributor did not want to squash commits (his own commits) so this was actually specifically added for that case. If I remove it, then we are not covering that case anymore, which was a big motivation for this PR.
If you'd like I can add "this does not apply when cherry-picking commits from others" in the text.

>Yes, and now we are established another set of rules that interpose. How do you guarantee that people will make this very clear? Should we need to update that documentation and say that everything here do not apply?

I can't guarantee it. But with this written down, I am confident we will have fewer conflicts.

>Well, this is why I'm a little afraid about this particular change:

This PR still has to go through the Process WG no matter what, so I don't think there's any risk we can rush it.

Reply:
>The copyright is apache, there is nothing else needed. If someone wants to add their name to the copyright list at the top of the file, they can do that. A commit does not imply copyright

I think @nandojve referred to the fact that the one thing is the license (i.e. Apache v2) and another thing the copyright assignments. I have no idea whether a commit authorship implies copyright, but as you say the author can always add themselves to the top of the file so that they don't need to rely on version control to ensure their rights are respected.

>Again, you don't know what you are talking about.

Let's try to keep the conversation constructive.

Reply:
> If you'd like I can add "this does not apply when cherry-picking commits from others" in the text.

Absolutely against this, as said authors can add their copyright to things or they can make self contained edits, we don't need a sprawl of commits just to do one thing

Reply:
@carlescufi ,

> Again, these are guidelines, not a precise rulebook nor a contract. 

Then we need to think about the `must` word in the text. I'll make a suggestion:

```diff
- the git history during development of a pull request must be squashed out of these
- fundamental commits whenever bisectability allows it.
+ the git history during development of a pull request should be squashed out of these
+ fundamental commits whenever bisectability allows it and the change do not violate distinguishing
+ contributions from multiple authors.
```

This means:

Author A write a commit with: "This is my new contribution." and
Another Author B write another commit with: "And I continue the work."
The project must not obligate a squash like: This is my new contribution. And I continue the work." One Author 2 sign off.

It is necessary to keep 2 separated commits IF authors want to keep the distinct contributions. Reviewers can ask to squash but should never enforce it.

Is this make sense?

Reply:
That sounds reasonable, yes. Let me try to squeeze it in, thank you!

Reply:
>Absolutely against this, as said authors can add their copyright to things or they can make self contained edits, we don't need a sprawl of commits just to do one thing

@nordicjm I think the proposal by @nandojve is reasonable enough. It leaves the door open to collaboration in the case where multiple commit authors are involved, which is a sticking point by many clearly.

Reply:
> which is a sticking point by many clearly.

One person? I don't see anyone else having a problem with it, copyright has nothing to do with who authored a commit and if 3 people contribute to e.g. a board and it makes it a mess as per the original PR in question it should be squashed, there is no reason at all it should not be left as 3 commits. Another example is hwmv2, that is squashed to a single commit, that doesn't mean one person owns the copyright or made the changes to over 600 boards and their SoCs and no-one has a problem with that

Reply:
@carlescufi 
> Again, these are guidelines, not a precise rulebook nor a contract.

An open question/exercise: if it's permitted/tacitly endorsed to squashes commits authored by a different contributor(s), how is the "Squasher" going to fulfill the "prominent notice" requirement of `4b` in the Apache 2.0 Licence? 

I suggest that the easiest way to store/show "prominent notice" is for the committed lines of code to have the correct (git) Author. Squashed commits from >1 Authors (Alice and Bob) with a distinct Committer (Clare) is then not permitted, because commits only have one Committer (Clare) and one Author (Alice, or Bob). 

Echoing @nandojve 's excellent points above, we need to be very mindful of the terms of the Apache 2.0 License. The guidelines aren't a contract, but they do... guide... everyone on how to meet the obligations of the License. Section 4 needs consideration. 


> The copyright is apache, there is nothing else needed.

This statement is not correct, and the thought-terminating ending is concerning. @jhedberg has explained this very well at https://github.com/zephyrproject-rtos/zephyr/pull/83117#discussion_r1898508070 . Maintainers, please acknowledge and thank people for correcting mistakes/misunderstandings ("whoops, well spotted"/"you're right, my bad"/"You are right, my mistake"). This would improve credibility and show 'good faith' in conversation.

Reply:
>An open question/exercise: if it's permitted/tacitly endorsed to squashes commits authored by a different contributor(s), how is the "Squasher" going to fulfill the "prominent notice" requirement of 4b in the Apache 2.0 Licence?

You mean "You must cause any modified files to carry prominent notices stating that You changed the files'? This is already not being respected in a literal sense in many of the PRs today, because it is extremely common to make changes to files without modifying the copyright list at the beginning of the file, unless I misunderstand you.

>I suggest that the easiest way to store/show "prominent notice" is for the committed lines of code to have the correct (git) Author. Squashed commits from >1 Authors (Alice and Bob) with a distinct Committer (Clare) is then not permitted, because commits only have one Committer (Clare) and one Author (Alice, or Bob).

The license says the notice should be carried by the files themselves, not by version control metadata. Given that we are not complying with that in the first place, I think just using `Co-authored-by` or explaining with text in the commit message is likely (IANAL) legally equivalent to what you propose.

>Maintainers, please acknowledge and thank people for correcting mistakes/misunderstandings ("whoops, well spotted"/"you're right, my bad"/"You are right, my mistake"). This would improve credibility and show 'good faith' in conversation.

Completely agree with you here, I always try to do that but may not always remember. It's good to be reminded.